lib/srv/bpf_test.go

```diff
@@ -0,0 +1,850 @@
+	args []string
+	// interpreter for the command, if the command is a script
+	interpreter string
+	// path to the script that will be executed
+	scriptPath string
+	// should be set to true if the command is expected to return with
+	// a non-zero exit code
+	expectedFail bool
+}
+
+type addrInfo struct {
+	// destination address the command will connect to
+	addr string
+	// destination port the command will connect to
+	port int
+}
+
+// TestBPFRecording runs various commands with Enhanced Session Recording
+// enabled and verifies that the recorded events are expected.
+func TestBPFRecording(t *testing.T) {
+	checkBPF(t)
+
+	srv := newMockServer(t)
+	bpfSrv := newBPFService(t)
+
+	// Create a temp dir and files for commands to use.
+	cmdDir := t.TempDir()
+	tempFilePath := filepath.Join(cmdDir, "file")
+	tempFile, err := os.Create(tempFilePath)
+	require.NoError(t, err)
+	require.NoError(t, tempFile.Close())
+	newFilePath := filepath.Join(cmdDir, "newfile")
+
+	// Lookup paths of programs that are also shell builtins.
+	echoPath, err := exec.LookPath("echo")
+	require.NoError(t, err, "echo command is required for these tests but was not found")
+
+	// Create a TCP listener for each IP family.
+	lis4, err := net.Listen("tcp4", "127.0.0.1:0")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		lis4.Close()
+	})
+	netAddr4 := lis4.Addr().String()
+	addr4, port4Str, err := net.SplitHostPort(netAddr4)
+	require.NoError(t, err)
+	port4, err := strconv.Atoi(port4Str)
+	require.NoError(t, err)
+
+	lis6, err := net.Listen("tcp6", "[::1]:0")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		lis6.Close()
+	})
+	netAddr6 := lis6.Addr().String()
+	addr6, port6Str, err := net.SplitHostPort(netAddr6)
+	require.NoError(t, err)
+	port6, err := strconv.Atoi(port6Str)
+	require.NoError(t, err)
+
+	var wg sync.WaitGroup
+	wg.Go(func() { go handleConnections(lis4) })
+	wg.Go(func() { go handleConnections(lis6) })
+	t.Cleanup(wg.Wait)
+
+	// Create obfuscated script to run.
+	obfScript := fmt.Sprintf(`#!/bin/bash
+eval $(echo %s | base64 --decode)`,
+		base64.StdEncoding.EncodeToString([]byte("ls -h")),
+	)
+	obfScriptName := "obf.sh"
+	obfScriptPath := filepath.Join(cmdDir, obfScriptName)
+	err = os.WriteFile(obfScriptPath, []byte(obfScript), 0o700)
+	require.NoError(t, err)
+
+	// Create a slice of arguments over the maximum length.
+	overMaxArgs := make([]string, maxArgs+3)
+	for i := range overMaxArgs {
+		overMaxArgs[i] = strconv.Itoa(i) + overMaxArg
+	}
+	atMaxArgs := slices.Clone(overMaxArgs)
+	for i := range atMaxArgs {
+		atMaxArgs[i] = atMaxArgs[i][:maxArgLength]
+	}
+	maxArgPaths := slices.Clone(atMaxArgs)
+	for i := range maxArgPaths {
+		maxArgPaths[i] = maxArgPaths[i][:maxPathLength]
+	}
+
+	// Define the test cases.
+	tests := []struct {
+		name       string
+		command    string
+		eventInfos []expectedEvents
+	}{
+		{
+			name:    "no commands",
+			command: "true",
+		},
+		{
+			name:    "basic command",
+			command: fmt.Sprintf("find %s", cmdDir),
```

Comment:
Suggestion: Here and below prefer string concatenation over fmt.Sprintf for simple cases. See https://google.github.io/styleguide/go/best-practices#string-concatenation.
```suggestion
			command: "find " + cmdDir,
```